CRM/Airmail/Backend/Elastic.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+use CRM_Airmail_Utils as E;
+
+class CRM_Airmail_Backend_Elastic implements CRM_Airmail_Backend {
+
+  public function processInput($input) {
+    return $input;
+  }
+
+  public function validateMessages($event) {
+    return !empty($event['transaction']);
+  }
+
+  public function processMessages($event) {
+    $status = $event['status'];
+    //Parse postback url to get all required mailing information.
+    $mailingJobInfo = E::parseSourceString($event['postback']);
+    //Map Elastic email bounce types and resoan with CiviCRM.
+    $bounce_details = self::getBounceTypeMessages($event);
+    $params = [
+      'job_id' => $mailingJobInfo['job_id'],
+      'event_queue_id' => $mailingJobInfo['event_queue_id'],
+      'hash' => $mailingJobInfo['hash'],
+      'bounce_type_id' => $bounce_details['bounce_type_id'],
+      'bounce_reason' => $bounce_details['bounce_reason'],
+    ];
+    switch ($status) {
+      // When you want to receive notifications for bounced emails.
+      case 'Bounce / Error':
+      case 'Bounce':
+      case 'Error':
+        // Add bounce report in CiviCRM.
+        CRM_Airmail_EventAction::bounce($params);
+        break;
+    }
+  }
+
+  /**
+   * Called by hook_civicrm_alterMailParams
+   *
+   * @param array $params
+   *   The mailing params
+   * @param string $context
+   *   The mailing context.
+   */
+  public function alterMailParams(&$params, $context) {
+    // Add custom headers for ElasticEmail
+    // This is required so we will get postback URL in Elastic email response,
+    // and can then identify the mailing item, source contact, etc.
+    if ($context != 'messageTemplate') {
+      if (!array_key_exists('headers', $params)) $params['headers'] = array();
+      $params['headers']['X-ElasticEmail-Postback'] = $params['Return-Path'];
+    }
+  }
+
+ /**
+  *
+  * Function for getting bounce type and message.
+  */
+  public function getBounceTypeMessages($event) {
+    //Elastic Email bounce categories.
+    $elastic_categories = array(
+      'Away' => array('Unknown'),
```

Comment:
Does this mean unknown bounce types are treated as Away and the email will only be put on hold after 30 bounces? Might this be too optimistic and bad for the sender reputation?

Is there any way to filter out out of office messages?

Reply:
I would argue it make sense to stop sending to an email address after 30 emails do soft bounce, whatever the reason is.